network/newstream/stream.go

```diff
@@ -17,151 +17,1108 @@
-	defer s.removePeer(sp)
-	return peer.Run(sp.HandleMsg)
+	if err := p.Send(ctx, wantedHashesMsg); err != nil {
+		p.logger.Error("error sending wanted hashes", "err", err)
+		p.Drop()
+		return
+	}
+	if ctr == 0 {
+		// request the next range in case no chunks wanted
+		r.requestSubsequentRange(ctx, p, provider, w, msg.LastIndex)
+		return
+	}
+	select {
+	case err := <-errc:
+		if err != nil {
+			streamBatchFail.Inc(1)
+			p.logger.Error("got an error while sealing batch", "from", w.from, "to", w.to, "err", err)
+			p.Drop()
+			return
+		}
+
+		// seal the interval
+		if err := p.sealWant(w); err != nil {
+			p.logger.Error("error persisting interval", "from", w.from, "to", w.to, "err", err)
+			p.Drop()
+			return
+		}
+	case <-time.After(activeBatchTimeout):
+		p.logger.Error("batch has timed out", "ruid", w.ruid)
+		close(w.closeC) // signal the polling goroutine to terminate
+		p.mtx.Lock()
+		delete(p.openWants, msg.Ruid)
+		p.mtx.Unlock()
+		p.Drop()
+		return
+	case <-r.quit:
+		return
+	case <-p.quit:
+		return
+	}
+	r.requestSubsequentRange(ctx, p, provider, w, msg.LastIndex)
+}
+
+// serverHandleWantedHashes is handled on the server side (Peer is the client) and is dependent on a preceding OfferedHashes message
+// the method is to ensure that all chunks in the requested batch is sent to the client
+func (r *Registry) serverHandleWantedHashes(ctx context.Context, p *Peer, msg *WantedHashes, o offer, provider StreamProvider) {
+	p.logger.Debug("serverHandleWantedHashes", "ruid", msg.Ruid)
+	start := time.Now()
+	defer func(start time.Time) {
+		metrics.GetOrRegisterResettingTimer("network.stream.handle_wanted_hashes.total-time", nil).UpdateSince(start)
+	}(start)
+
+	defer func() {
+		p.mtx.Lock()
+		delete(p.openOffers, msg.Ruid)
+		p.mtx.Unlock()
+	}()
+
+	var (
+		l          = len(o.hashes) / HashSize
+		cd         = &ChunkDelivery{Ruid: msg.Ruid}
+		wantHashes = []chunk.Address{}
+		allHashes  = make([]chunk.Address, l)
+	)
+
+	if len(msg.BitVector) == 0 {
+		p.logger.Debug("peer does not want any hashes in this range", "ruid", o.ruid)
+		for i := 0; i < l; i++ {
+			allHashes = append(allHashes, o.hashes[i*HashSize:(i+1)*HashSize])
+		}
+		// set all chunks as synced
+		if err := provider.Set(ctx, allHashes...); err != nil {
+			p.logger.Error("error setting chunk as synced", "addrs", allHashes, "err", err)
+			p.Drop()
+			return
+		}
+		return
+	}
+	want, err := bv.NewFromBytes(msg.BitVector, l)
+	if err != nil {
+		p.logger.Error("error initiaising bitvector", "l", l, "ll", len(o.hashes), "err", err)
+		p.Drop()
+		return
+	}
+
+	maxFrame := MinFrameSize
+	if v := BatchSize / 4; v > maxFrame {
+		maxFrame = v
+	}
+
+	// check which hashes to get from the localstore
+	for i := 0; i < l; i++ {
+		hash := o.hashes[i*HashSize : (i+1)*HashSize]
+		if want.Get(i) {
+			metrics.GetOrRegisterCounter("network.stream.handle_wanted.want_get", nil).Inc(1)
+			wantHashes = append(wantHashes, hash)
+		}
+		allHashes[i] = hash
```

Comment:
All hashes are needed for mode set sync. If the peer does not want some hashes, it means that it already has them. Until we have a confirmation protocol that proves that, we need to set all hashes to sync state so that they can be garbage collected.